### PR TITLE
404 describe

### LINKED ISSUE #3092: "Error: 404 page not found" when querying an older Zed lake service
Repro is with Zui commit 21abbf6. This issue was reported by a community zync user. They connect to remote lakes in their environment that are running Zed lake service `v1.14.0`.

We prepare by starting a Zed lake `v1.14.0` service and loading it with some data.

```
$ zed serve -lake lake -l localhost:8888
```

```
$ zed -lake http://localhost:8888 create -use foo
$ zed -lake http://localhost:8888 load sample.zng
```

In the attached video, we first make these preparations, then from within Zui 21abbf6 we click **Add Lake** and make the connection to the "remote" service we just started on `localhost:8888`. For any query we issue, instead of query results we are shown **Error: 404 page not found**.

https://github.com/brimdata/zui/assets/5934157/6dfe6729-ff96-4346-a870-9916fc8abd18

We discussed this one as a group. The root cause of the problem begins with the changes in #3076 when Zui began hitting the Zed service `/query/describe` endpoint introduced in https://github.com/brimdata/zed/pull/5126. Interestingly, the user-facing symptom did _not_ surface right at the point when #3076 merged. Studying traffic in Wireshark we can see that the HTTP 404 is being returned by the Zed service when the app tries to hit `/query/describe`, but the app was not surfacing it and instead the user saw their query result as always. It's only with the changes to the error handling in #3085 that the **Error: 404 page not found** started being surfaced.

In terms of a fix, we agreed the high-level goal is that we want Zui at tip of `main` to still be able to work with lake service versions that lack the `/query/describe` endpoint, even if some functionality is not available due to the version mismatch. We reached consensus that a reasonable fix would be for the app to recognize if `/query/describe` is not available and, if so, just execute the query and present any errors that come back. This means that the user will be missing out on the improved error handling we're enabling via `/query/describe` so we could perhaps surface some informational message to that effort (i.e., to nudge them toward upgrading their lake service versions) but this was not seen as "must have" functionality. It was also noted that pagination and drill-down would not work in the absence of `/query/describe`, so we should make sure this functionality gets disabled in a friendly way.

### LINKED ISSUE #3100: Crash when accessing remote lake (broke at 2473666)
Repro is with Zui commit 2473666.

In the attached video, I click **Add Lake** in an attempt to connect to a remote lake, specifically the one behind my Zui Insiders at `http://localhost:9988`. As the video shows, this results in a crash and a failure to access the lake.

https://github.com/brimdata/zui/assets/5934157/8f78e472-ddbb-4601-b08a-f5e0fac0390d

The complete error dump:

```
Unhandled Runtime Error
Error: <!DOCTYPE html><html><head><style data-next-hide-fouc="true">body{display:none}</style><noscript data-next-hide-fouc="true"><style>body{display:block}</style></noscript><meta charSet="utf-8"/><meta name="viewport" content="width=device-width"/><title>404: This page could not be found</title><meta name="next-head-count" content="3"/><noscript data-n-css=""></noscript><script defer="" nomodule="" src="/_next/static/chunks/polyfills.js?ts=1718733506065"></script><script src="/_next/static/chunks/webpack.js?ts=1718733506065" defer=""></script><script src="/_next/static/chunks/main.js?ts=1718733506065" defer=""></script><script src="/_next/static/chunks/pages/_app.js?ts=1718733506065" defer=""></script><script src="/_next/static/chunks/pages/_error.js?ts=1718733506065" defer=""></script><script src="/_next/static/development/_buildManifest.js?ts=1718733506065" defer=""></script><script src="/_next/static/development/_ssgManifest.js?ts=1718733506065" defer=""></script><noscript id="__next_css__DO_NOT_USE__"></noscript></head><body><div id="__next"><div style="font-family:system-ui,&quot;Segoe UI&quot;,Roboto,Helvetica,Arial,sans-serif,&quot;Apple Color Emoji&quot;,&quot;Segoe UI Emoji&quot;;height:100vh;text-align:center;display:flex;flex-direction:column;align-items:center;justify-content:center"><div><style>body{color:#000;background:#fff;margin:0}.next-error-h1{border-right:1px solid rgba(0,0,0,.3)}@media (prefers-color-scheme:dark){body{color:#fff;background:#000}.next-error-h1{border-right:1px solid rgba(255,255,255,.3)}}</style><h1 class="next-error-h1" style="display:inline-block;margin:0 20px 0 0;padding-right:23px;font-size:24px;font-weight:500;vertical-align:top;line-height:49px">404<!-- --></h1><div style="display:inline-block;text-align:left"><h2 style="font-size:14px;font-weight:400;line-height:49px;margin:0">This page could not be found<!-- -->.<!-- --></h2></div></div></div></div><script src="/_next/static/chunks/react-refresh.js?ts=1718733506065"></script><script id="__NEXT_DATA__" type="application/json">{"props":{"pageProps":{"statusCode":404}},"page":"/_error","query":{"ctrl":"true"},"buildId":"development","isFallback":false,"gip":true,"scriptLoader":[]}</script></body></html>

Source
../packages/zed-js/src/util/error.ts (15:30) @ createError

  13 |   return e;
  14 | }
> 15 | if (isString(input)) return new Error(input);
     |                            ^
  16 | 
  17 | return new Error('Unknown Error');
  18 | }
Call Stack
Client.send
../packages/zed-js/src/client/base-client.ts (190:40)
async Client.query
../packages/zed-js/src/client/base-client.ts (41:19)
async Client.getPools
../packages/zed-js/src/client/base-client.ts (99:17)
```

This crash started occurring at this commit 2473666, which is associated with the changes in #3096.

### PR DESCRIPTION
**Pull request by jameskerr**
Fixes #3092
Fixes #3100

- [x] Disable Pagination (i.e., limited to 500 rows of query output in window)
- [x] Disable Drill Down (i.e., right-click **Pivot to Values**)
- [x] Disable Enhanced Error Reporting (e.g., red squiggly underlines for syntax errors in the editor)
- [x] Disable **Correlations** views in **Details** side panel & window

The app in this PR does a feature detection step whenever the lake changes or when it boots up for the first time. It will check if the describe endpoint returns a 404. If so, it will disable the following checkboxes above.

Everything seems to be working as intended in this.

### CODE REVIEW

File: apps/zui/src/app/lakes/root.tsx [hunk 1]
```diff
@@ -27,6 +27,7 @@ export function InitLake({children}) {
   useLayoutEffect(() => {
     if (status) return
     if (lake) dispatch(updateStatus(lake.id))
+    Active.lake.sync()
   }, [lake?.id, status])
```
Comment (jameskerr) [inline apps/zui/src/app/lakes/root.tsx:31]: Whenever the current lake ID changes, sync our info about the lake with the lake service.

File: apps/zui/src/app/lakes/root.tsx [hunk 2]
```diff
@@ -27,6 +27,7 @@ export function InitLake({children}) {
   useLayoutEffect(() => {
     if (status) return
     if (lake) dispatch(updateStatus(lake.id))
+    Active.lake.sync()
```
Comment (jameskerr) [inline apps/zui/src/app/lakes/root.tsx:30]: console.log('Clearing the timer');

File: apps/zui/src/app/lakes/root.tsx [hunk 3]
```diff
@@ -35,7 +36,7 @@ export function InitLake({children}) {
 
   switch (status) {
     case "disconnected":
-      return <ConnectionError onRetry={() => dispatch(initCurrentTab())} />
+      return <ConnectionError onRetry={() => dispatch(updateStatus(lake.id))} />
```
Comment (jameskerr) [inline apps/zui/src/app/lakes/root.tsx:39]: `initCurrentTab` only called this `updateStatus` function. So I inlined it and removed `initCurrentTab`. 

All of this "lake" status/authentication code needs another look at some point.

File: apps/zui/src/js/initializers/init-lake.ts [hunk 4]
```diff
@@ -2,10 +2,13 @@ import Current from "../state/Current"
...
 import {defaultLake} from "./initLakeParams"
 import Window from "../state/Window"
+import {Lake} from "src/models/lake"
 
 export function initLake(store: Store) {
   const lakeId = Current.getLakeId(store.getState())
   const id = lakeId || defaultLake().id
   // This sets up the lake's tabs
+  const lake = Lake.find(id)
+  lake.sync()
```
Comment (jameskerr) [inline apps/zui/src/js/initializers/init-lake.ts:12]: As soon as the app boots up, we sync the lake.

File: apps/zui/src/js/initializers/initialize.ts [hunk 5]
```diff
@@ -44,11 +44,7 @@ export default async function initialize(
   const api = new ZuiApi()
   const store = await initStore(api, renderer)
   const asyncTasks = initAsyncTasks(renderer)
-  await initGlobals(store)
-  await initLake(store)
-  api.init(store.dispatch, store.getState)
-  initDOM()
-  initIpcListeners(store)
+  initDomainModels({store})
```
Comment (jameskerr) [inline apps/zui/src/js/initializers/initialize.ts:47]: the Domain Models need to be provisioned with the store variable before we call "initLake". So I re-ordered the steps here.

File: apps/zui/src/js/state/Lakes/types.ts [hunk 6]
```diff
@@ -6,6 +6,9 @@ export type LakeAttrs = {
   version?: string
   authType: AuthType
   authData?: AuthData
+  features?: {
+    describe: boolean
+  }
```
Comment (jameskerr) [inline apps/zui/src/js/state/Lakes/types.ts:11]: I'm going to add a "features" property the the data we cache about the lake. The only feature right now will be "describe".

File: apps/zui/src/models/active.ts [hunk 7]
```diff
@@ -36,4 +37,9 @@ export class Active extends DomainModel {
       id: globalThis.windowId,
     })
   }
+
+  static get lake() {
+    const id = this.select(Current.getLakeId)
+    return Lake.find(id)
+  }
```
Comment (jameskerr) [inline apps/zui/src/models/active.ts:44]: This is a shortcut to get the Active lake object. `Active.lake`

File: apps/zui/src/models/lake.ts [hunk 8]
```diff
@@ -1,7 +1,20 @@
...
 import {LakeAttrs} from "../js/state/Lakes/types"
+import {Client} from "@brimdata/zed-js"
+import Slice from "src/js/state/Lakes"
+import LakeStatuses from "src/js/state/LakeStatuses"
+import {LakeStatus} from "src/js/state/LakeStatuses/types"
+import {FeatureDetector} from "./lake/feature-detector"
+
+type LakeModelAttrs = LakeAttrs & {status: LakeStatus}
+
+export class Lake extends DomainModel<LakeModelAttrs> {
```
Comment (jameskerr) [inline apps/zui/src/models/lake.ts:11]: Here we are adding the methods we need to the Lake model. 

I've added:

- find(id)
- client 
- features
- sync()
- update
- save
- isRunning()

File: packages/zed-js/src/client/base-client.ts [hunk 9]
```diff
@@ -52,22 +52,24 @@ export abstract class BaseClient {
...
       path: `/query/describe`,
       body: JSON.stringify({ query, head }),
       contentType: 'application/json',
       format: 'json',
       signal: abortCtl.signal,
       timeout: options.timeout,
-      dontRejectError: true,
     });
-    return result.json();
+    if (response.ok || response.status === 400) {
```
Comment (jameskerr) [inline packages/zed-js/src/client/base-client.ts:68]: Re-arranged the code to allow for an error response from the server. If the describe endpoint returns code 400, there is a problem with the query, but we don't want to throw an error. We want to report the problem to the user. So we parse the JSON as an expected response.

File: packages/zed-js/src/client/base-client.ts [hunk 10]
```diff
@@ -145,18 +147,7 @@ export abstract class BaseClient {
...
-    body?: Types.IsoBody;
-    format?: Types.ResponseFormat;
-    signal?: Types.IsoAbortSignal;
-    headers?: Record<string, string>;
-    timeout?: number;
-    contentType?: string;
-    duplex?: 'half';
-    dontRejectError?: boolean;
-  }) {
+  protected request(opts: Types.RequestOpts) {
```
Comment (jameskerr) [inline packages/zed-js/src/client/base-client.ts:150]: I separated the creation of the request into this request method. The existing `send` method then calls this.

The `describeQuery` method only uses `request` and handles errors in a different way than the other methods.

**Comment by philrz** [comment]
I'm currently testing this branch at commit 31b5d55.

While the changes in this PR are primarily concerned with connecting to older Zed lakes, I happened to notice something unexpected when connecting to a _current_ Zed lake. In the attached video I'm connecting to the lake running behind Zui Insiders `1.7.1-insiders.4`, so that's Zed commit `d217ed8` at the moment. When I click on a Zeek `conn` record, I noticed that the **Correlation** view in the **Detail** panel is not populated and the following error message appears in my console where I launched the branch Zui 31b5d55 in dev mode.

```
16:13:10.895 › Failed to compile query for zui-zeek/uid-correlation
16:13:10.895 › invalid operation: null: pool not found at line 2, column 8:
  from null
       ~~~~
    at createError (/Users/phil/work/zui/apps/zui/dist/main.js:69495:30)
    at ElectronZedClient.send (/Users/phil/work/zui/apps/zui/dist/main.js:69756:29)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
    at async ElectronZedClient.query (/Users/phil/work/zui/apps/zui/dist/main.js:69633:21)
    at async CorrelationApi.query (/Users/phil/work/zui/apps/zui/dist/main.js:75570:19)
    at async CorrelationsApi.compile (/Users/phil/work/zui/apps/zui/dist/main.js:67280:26)
    at async WebContents.<anonymous> (node:electron/js2c/browser_init:2:82807)
```

https://github.com/brimdata/zui/assets/5934157/82fc1e98-a7ec-4e6c-bbd1-f3b7f5fe179a

I went back to Zui commit 1eac49a which was right before we broke the ability to connect to remote lakes (#3100) and this was working back then.

**Comment by philrz** [comment]
I'm currently testing this branch at commit 62a6084.

The attached video shows a successful walk-through of the functionality that these changes disable when accessing an older Zed lake service. As prep, I have the [ZNG format Zed sample data](https://github.com/brimdata/zed-sample-data/tree/main/zng) loaded to both the lake running behind Zui (Zed commit `13d15d3`) and a separate "remote" Zed `v1.14.0` lake I've started on port `8888`. As a baseline I first demo each of the four functional areas described by the checkboxes in the opening PR text while accessing the data in the local lake, then I access the `v.14.0` lake and confirm that each of these is disabled.

https://github.com/brimdata/zui/assets/5934157/b215f4b3-ca2d-40b5-939e-7f2f35905433

**Review by philrz — APPROVED** [review]

**Pull request merged**